src/NServiceBus.Transport.RabbitMQ/Connection/ChannelProvider.cs

```diff
@@ -84,8 +93,15 @@ public void ReturnPublishChannel(ConfirmsAwareChannel channel)
             }
         }
 
-        public void Dispose()
+#pragma warning disable PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        public async ValueTask DisposeAsync()
+#pragma warning restore PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
         {
+            stoppingTokenSource.Cancel();
+            stoppingTokenSource.Dispose();
+
+            await reconnectTask.ConfigureAwait(false);
```

Comment:
I don't see there being any value in trying to wait for a possible reconnect loop here. Removing this would simplify things and remove some of the other changes in this PR.

Reply:
I brought up the same thing originally. Because from a shutdown perspective I agree. The problem is though if you don't await it can happen due to the asynchronous nature we race to create a new one while we try to dispose. All possible permutations of things can happen here while probably the most likely outcome will be things are disposed. This would make the shutdown deterministic

Reply:
And also problematic if one endpoint that didn't throw on stop is still holding on remote resources while another endpoint instance is spun up which is dependent on those same resources and rejected by the broker as it reached its connection limit.

If a user isn't interested in that it should simply not await and do `_ = endpointInstance.Stop();` but users should be able to assume that after `.Stop()` returns all resources are released.